### ISSUE TITLE
Rendering & platform support

### ISSUE THREAD

**Issue by itsezc** [DioxusLabs/blitz#25]
From reading the Dioxus docs and any resources I could come across (on Discord and the wider web), I understand that blitz (plans to) support the following platforms:

- Desktop (Windows, Linux and MacOS)
- Mobile (Android & iOS)
- Bevy (among other game engines)

While this is an amazing list, I would be interested to know the following (for my own education):

- Is the rendering immediate or retained?
- Is there any planned support for the web via webgpu a potential case use case would be building GUI heavy apps i.e. something similar to Unreal Engine's blueprint editor in Dioxus?
- Given that web would be a supported platform via blitz, how would it address SEO related issues (if at all, even if it were through Dioxus)
- Any intentions on targetting embedded platforms i.e. Raspberry Pi and some other / more niche platforms used in IoT where GUIs are present

I'm not an expert in renderers or vDOMs so excuse my ignorance, this issue is for me (and people alike) to get a better understanding of blitz and the role it'll play.

**Comment by Demonthos** [comment]
> From reading the Dioxus docs and any resources I could come across (on Discord and the wider web), I understand that blitz (plans to) support the following platforms:
> 
> * Desktop (Windows, Linux and MacOS)
> * Mobile (Android & iOS)
> * Bevy (among other game engines)

I would be open to adding Bevy support, but currently it is not a priority.

> While this is an amazing list, I would be interested to know the following (for my own education):
> 
> * Is the rendering immediate or retained?

The whole screen is redrawn every frame, but the plan is to switch to partial screen updates at some point. The Virtual DOM (or [your own framework](https://github.com/DioxusLabs/blitz/pull/22)) creates a retained tree of elements. From the elements Blitz creates a retained tree of computed styles and layout. When the styles and layout are rendered the whole screen is redrawn.

> * Is there any planned support for the web via webgpu a potential case use case would be building GUI heavy apps i.e. something similar to Unreal Engine's blueprint editor in Dioxus?

There is no plans to support web though Blitz, except potentially as a demo platform. For web rendering, you can use the Dioxus-web crate. This is the main reason Blitz renders normal HTML elements + attributes. HTML elements have better out of the box support for accessibility and SEO optimization. It should be possible to target web through native elements and Vello at the same time. This is also not supported by [Vello](https://github.com/linebender/vello#web).

> * Given that web would be a supported platform via blitz, how would it address SEO related issues (if at all, even if it were through Dioxus)

You can use Dioxus SSR to make Dioxus sites work with SEO.

> * Any intentions on targeting embedded platforms i.e. Raspberry Pi and some other / more niche platforms used in IoT where GUIs are present

There are no plans for `no-std` support, but targeting embedded platforms that are supported by rust could be interesting in the future. This again is limited by platforms Vello supports. Currently the focus is on building a stable platform for native applications on desktop and mobile.

**Comment by itsezc** [comment]
Thanks for that provides a much greater insight into the ecosystem, though I'd like to advocate for definitely adding web as one of the primary targets (that is with WebGPU and not just DOM) for when you want to build the next Figma so I can see the benefit of having both rendering strategies.

@Demonthos is there a comparison sheet that has been compiled to view the difference between Blitz and other rendering engines such as UltraLight, WRY etc. and also perhaps the browsers so we can use that to track what features are available?

**Comment by badlanguage** [comment]
For multi platform support is root font size reset supported with Blitz?

```css

:root {
	font-size: 62.5%;
}

body {
	font-size: 1.6rem;
}
```

So I can use rem as units of 10

@Demonthos

**Comment by Demonthos** [comment]
> @Demonthos is there a comparison sheet that has been compiled to view the difference between Blitz and other rendering engines such as UltraLight, WRY etc. and also perhaps the browsers so we can use that to track what features are available?

I have a private tracking board with what Blitz supports. I'm working on making it public soon.

Currently, Blitz supports a tiny subset of the attribute/elements that you would find in a normal browser, but we are looking to expand these in the future. Right now, Blitz is only really usable for small demos. In the future it will support a larger subset of HTML, but it will likely never support the entire spec.

> For multi platform support is root font size reset supported with Blitz?

Selectors are not implemented yet, but they are one of the things (along with https://github.com/DioxusLabs/dioxus/pull/940) that are required for widgets in Blitz, so they are planned.

**Comment by itsezc** [comment]
Thanks for opening up the tracking board and expanding on the topic.

> In the future it will support a larger subset of HTML, but it will likely never support the entire spec.

Is there a list of features / elements that Blitz won't support, and the reasoning? I'm interested to know the what and why and see if there are any limitations for my projects.

**Comment by Demonthos** [comment]
> Is there a list of features / elements that Blitz won't support, and the reasoning? I'm interested to know the what and why and see if there are any limitations for my projects.

There isn't a set list of items that will not be implemented. As we implement more elements and attributes I expect there will be areas that will are both:
1) Very rairly used in modern websites
2) Have substantial performance or maintainability impacts for much more common use cases

**Comment by nicoburns** [comment]
Some examples we could take inspiration from when producing a feature support table:

- Dropflow (https://github.com/chearon/dropflow#supported-css-rules)
- Ladybird (https://css.tobyase.de/)
- Sciter (https://sciter.com/docs/content/css/cssmap.html)
- litehtml (https://docs.google.com/spreadsheets/d/1CM6p2BU0XwM7KS0ZUIP7u25g2rrHXKx_ANGdDG98vwc)

**Comment by Maikel1963** [comment]
Hello 
I was happily impressed that I could run `cargo run --example todomvc --release` in cloned repository (I have to use target `windows_x86_64_gnu`)  but now stuck by trying to get this example work as standalone app. I guess there is some dependency missing in the `cargo.toml` but I do not understand, what.
Here is the content of my `cargo.toml`:

```
[dependencies]
dioxus = { git = "https://github.com/DioxusLabs/dioxus", features = [] }
dioxus-native = { git ="https://github.com/DioxusLabs/blitz" }
tracing-subscriber = "0.3.19"

[features]
default = ["desktop"]
web = ["dioxus/web"]
desktop = ["dioxus/desktop"]
mobile = ["dioxus/mobile"]
```
The error message I get is:

```
error[E0308]: mismatched types
  --> src\main.rs:8:25
   |
8  |   dioxus_native::launch(app)
   |   --------------------- ^^^ expected fn pointer, found fn item
   |   |
   |   arguments to this function are incorrect
   |
   = note: expected fn pointer `fn() -> Result<dioxus_core::nodes::VNode, dioxus_core::render_error::RenderError>`
                 found fn item `fn() -> Result<VNode, dioxus::prelude::RenderError> {app}`
note: function defined here
  --> C:\Users\michael.heisler\.cargo\git\checkouts\blitz-66635cc3152d32bd\0e53be5\packages\dioxus-native\src\lib.rs:29:8
   |
29 | pub fn launch(root: fn() -> Element) {
   |        ^^^^^^
```
What has to be different?

**Comment by nicoburns** [comment]
@Maikel1963 Try using `dioxus-native` from the Dioxus repo, `dioxus::launch` and adding a `native = ["dioxus/native"]` feature, and running with `dx serve --platform android --features native` (you may need to compile CLI from dioxus `main`).

(it is currently also possible to use the `dioxus-native` from the Blitz repo, to do that you would need dioxus `0.6` - but this is an old version that will be deleted soon).

**Comment by nicoburns** [comment]
> I have to use target `windows_x86_64_gnu`

Are you saying that the MSVC target doesn't work? What error are you getting there?

**Comment by Maikel1963** [comment]
No, it would work, but I am not allowed to use it because of licensing reasons. And with Blitz it was the first time that desktop is working without webview2loader.dll

**Comment by Maikel1963** [comment]
> Try using `dioxus-native` from the Dioxus repo, ...

This is definitively a step in the right direction. Now my `Cargo.toml` looks like this:

```
[dependencies]
dioxus = { git = "https://github.com/DioxusLabs/dioxus", features = [] }
#dioxus-native = { git ="https://github.com/DioxusLabs/blitz" }
tracing-subscriber = "0.3.19"

[features]
default = ["desktop"]
desktop = ["dioxus/desktop"]
native = ["dioxus/native"]
```
and the beginning of the `main.rs`:
```
use dioxus::prelude::*;
//use dioxus_native;
use std::collections::HashMap;

fn main() {
  tracing_subscriber::fmt::init();
  dioxus::launch(app)
}
```

Thus I can run it with `cargo run --features native`
But when trying to execute the exe file directly (doesn't matter if debug or release) I get this screen:

![Image](https://github.com/user-attachments/assets/a02e885b-f51a-4909-a1d4-9addeead3385)

What do I have to use to prevent the use of webview2loader.dll or how  is it integrated when running with cargo?
